tpm/signer.go

```diff
@@ -90,3 +91,41 @@ func (t *TPM) GetSigner(ctx context.Context, name string) (csigner crypto.Signer
 
 	return
 }
+
+// tss2Signer is a wrapper on top of [*tss2.Signer] that opens and closes the
+// tpm on each sign call.
+type tss2Signer struct {
+	*tss2.Signer
+	tpm *TPM
+}
+
+func (s *tss2Signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	ctx := context.Background()
+	if err = s.tpm.open(goTPMCall(ctx)); err != nil {
+		return nil, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer closeTPM(ctx, s.tpm, &err)
+	s.SetTPM(s.tpm.rwc)
+	signature, err = s.Signer.Sign(rand, digest, opts)
+	return
+}
+
+// GetTSS2Signer returns a crypto.Signer using the [tss2.TPMKey].
+func (t *TPM) GetTSS2Signer(ctx context.Context, key *tss2.TPMKey) (csigner crypto.Signer, err error) {
```

Comment:
I think this is not a great extension of the (public) API for the `tpm.TPM`. The reason is because it's now possible to provide a `tss2.TPMKey` that may or may not be (actually) backed by the `tpm.TPM` instance belonging to the TPM the TSS2 key was created for. Before this change, the `tpm.TPM` would be more of a facade to using the TPM, abstracting the underlying details such as key identifiers, storage format and attestation functionality. 

I understand that it's useful to get a reference to the TPMs `rwc`, but I think it's nicer to turn this dependency around? I think you'll still need it to be part of the `tpm` package, but it doesn't have to be a method on the `tpm.TPM`. Something like `Attach(key, tpm)` or `AttachTSS2(key, tpm)`, backed by roughly this method, but made private.

Another reason I think that's nicer is because we might still be able to implement a "full TSS2 format storage" that sits behind the `tpm.TPM` facade with all details abstracted.

Reply:
I don't get the name `Attach` why not just:
```go
func CreateTSS2Signer(ctx context.Context, t *TPM, key *tss2.TPMKey) (csigner crypto.Signer, err error)
```

Reply:
Changed to `CreateTSS2Signer` with 3f48274

Reply:
It's just about naming things, and about making clear what it does.

I thought of `Attach` because the `tss2.Key` "attaches" to the `tpm.TPM`'s `rwc`, which makes it a complete `crypto.Signer`. The `tss2.Key` is (mostly) something external to the `tpm.TPM` (currently), so it somehow needs to be connected to it. It's also sufficiently different from the other methods, because it's a bit of a special case. Hence `Attach`. 

`CreateTSS2Signer` at the package level is better than what it was, but I don't like that it makes `TSS2` part of the top level public API surface. To me, at the moment, it's more of a utility method required because of how the `tpm` package is structured. I would also consider it experimental atm.